File: skills/inspect-and-debug/harness-doctor/scripts/audit.py
#!/usr/bin/env python3
"""harness-doctor audit: fleet-wide mining of agent transcripts.

Mines the last N days of Claude Code transcripts under a projects root for
where wall-clock, tokens, and human corrections are wasted, and writes fact
tables plus a summary. Stdlib-only, Python 3.9+, cross-platform (no shelling
out to find/grep/du).

This is a mirror of audit.ts: identical CLI flags, identical camelCase fact
table field names, identical `--format json` report shape. The summary goes to
stdout, progress to stderr, and in `md` mode the last stdout line is the work
dir path unless `--no-tables` wrote nothing there. The fact tables contain real
prompts and command output; see SKILL.md.
"""

import argparse
import json
import os
import sys
import tempfile
from pathlib import Path

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from lib import aggregate, cost, render, scan  # noqa: E402

_TABLE_FILES = (
    ("sessions", "sessions.jsonl"),
    ("tools", "tools.jsonl"),
    ("bash", "bash.jsonl"),
    ("userMsgs", "user-msgs.jsonl"),
    ("errors", "errors.jsonl"),
)


def _build_parser():
    p = argparse.ArgumentParser(prog="audit.py", description="harness-doctor fleet audit")
    p.add_argument("--days", type=int, default=30, help="window in days, default 30")
    p.add_argument("--root", default=None, help="transcripts root, default <home>/.claude/projects")
    p.add_argument("--out", default=None, help="work dir for fact tables, default <tmpdir>/harness-audit")
    p.add_argument("--project", action="append", default=None,
                   help="only project dirs containing this substring (repeatable)")
    p.add_argument("--format", choices=["md", "json"], default="md", help="stdout summary format, default md")
    p.add_argument("--top", type=int, default=15, help="rows per ranking, default 15")
    p.add_argument("--no-tables", action="store_true", help="compute aggregates without writing the fact tables")
    p.add_argument("--no-redact", action="store_true", help="disable secret redaction of text fields")
    p.add_argument("--pricing", default=None, help="JSON file overriding the built-in per-model price table")
    return p


def _write_table(out_dir, name, rows):
    """Write one fact table as JSONL, in scan order."""
    path = out_dir / name
    with open(path, "w", encoding="utf-8") as fh:
        for row in rows:
            # Separators match JSON.stringify, so both runtimes write the same bytes.
            fh.write(json.dumps(row, ensure_ascii=False, separators=(",", ":")))
            fh.write("\n")
    return path


def main(argv=None):
    args = _build_parser().parse_args(argv)
    if args.days <= 0:
        print("--days needs a positive number, got %s" % args.days, file=sys.stderr)
        return 1

    root = Path(args.root) if args.root else Path.home() / ".claude" / "projects"
    out_dir = Path(args.out) if args.out else Path(tempfile.gettempdir()) / "harness-audit"
    redact_on = not args.no_redact

    try:
        pricing, overridden = cost.load_pricing(args.pricing)
    except (OSError, ValueError, json.JSONDecodeError) as exc:
        print("error: could not load --pricing file: %s" % exc, file=sys.stderr)
        return 1
    if overridden:
        print("using pricing override from %s" % args.pricing, file=sys.stderr)

    print("scanning %s, last %d days" % (root, args.days), file=sys.stderr)
    sessions, tools, bash_rows, user_msgs, errors, stats = scan.scan_all(
        root, args.days, args.project, redact_on, progress_stream=sys.stderr,
    )
    if stats["filesScanned"] == 0:
        print(
            "No transcripts in the last %d days under %s. "
            "Check --root, widen --days, or drop --project." % (args.days, root),
            file=sys.stderr,
        )

    tables = {"sessions": sessions, "tools": tools, "bash": bash_rows,
              "userMsgs": user_msgs, "errors": errors}
    if not args.no_tables:
        out_dir.mkdir(parents=True, exist_ok=True)
        for key, name in _TABLE_FILES:
            _write_table(out_dir, name, tables[key])
        print("wrote %d fact tables to %s" % (len(_TABLE_FILES), out_dir), file=sys.stderr)
        print(
            "  the tables hold real prompts and command output, redacted best-effort: review before sharing"
            if redact_on
            else "  WARNING: redaction DISABLED (--no-redact): the tables hold raw prompts and secrets",
            file=sys.stderr,
        )

    report = aggregate.build(
        sessions, tools, bash_rows, user_msgs, errors, stats, args.days, pricing, args.top,
    )

    if args.format == "json":
        print(render.render_json(report))
        return 0
    sys.stdout.write(render.render_md(report, args.top) + "\n")
    # The work dir is only worth printing when something was written there.
    if not args.no_tables:
        print(str(out_dir))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/inspect-and-debug/harness-doctor/scripts/audit.ts
#!/usr/bin/env -S node --experimental-strip-types
/**
 * harness-doctor: mine the last N days of agent transcripts for wasted
 * wall-clock, wasted tokens, and repeated human corrections.
 *
 * Runtime: zero build step, zero dependencies. Runs on Node >=22.18 (native TS
 * type-stripping) or Bun. On older Node use `npx tsx audit.ts ...`.
 *
 * Usage:
 *   node audit.ts [options]
 *   bun run audit.ts [options]
 *
 * Options:
 *   --days <n>          window in days, default 30
 *   --root <path>       transcripts root, default <home>/.claude/projects
 *   --out <dir>         work dir for fact tables, default <tmpdir>/harness-audit
 *   --project <substr>  only project dirs containing this substring (repeatable)
 *   --format md|json    stdout summary format, default md
 *   --top <n>           rows per ranking, default 15
 *   --no-tables         compute aggregates without writing the fact tables
 *   --no-redact         disable secret redaction of text fields (ON by default)
 *   --pricing <file>    JSON overriding the built-in per-model price table
 *
 * The summary goes to stdout, progress to stderr, and in `md` mode the last
 * stdout line is the work dir path unless `--no-tables` wrote nothing there.
 * The fact tables contain real prompts and command output: treat the work dir
 * as sensitive.
 */
import { readFileSync } from "node:fs";
import { homedir, tmpdir } from "node:os";
import { join } from "node:path";
import { scanTranscripts, writeTables } from "./lib/scan.ts";
import { aggregate } from "./lib/aggregate.ts";
import { renderMarkdown } from "./lib/render.ts";
import type { AggregateOptions, AuditOptions, PriceTable } from "./lib/types.ts";

const USAGE =
  "Usage: node audit.ts [--days n] [--root path] [--out dir] [--project substr]... " +
  "[--format md|json] [--top n] [--no-tables] [--no-redact] [--pricing file]";

/** Read a price-table override, exiting with a clear message when it is unusable. */
const loadPricing = (path: string): PriceTable => {
  try {
    const parsed = JSON.parse(readFileSync(path, "utf8")) as PriceTable;
    if (!parsed || typeof parsed !== "object") throw new Error("not an object");
    return parsed;
  } catch (err) {
    console.error(`Cannot read --pricing ${path}: ${err instanceof Error ? err.message : String(err)}`);
    process.exit(1);
  }
};

/** Parse argv into fully resolved options. Unknown flags are a hard error. */
export const parseArgs = (argv: string[]): AuditOptions => {
  const o: AuditOptions = {
    days: 30,
    root: join(homedir(), ".claude", "projects"),
    out: join(tmpdir(), "harness-audit"),
    projects: [],
    format: "md",
    top: 15,
    writeTables: true,
    redact: true,
  };
  const num = (raw: string | undefined, flag: string) => {
    const n = Number(raw);
    if (!Number.isFinite(n) || n <= 0) {
      console.error(`${flag} needs a positive number, got ${String(raw)}`);
      process.exit(1);
    }
    return n;
  };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i]!;
    if (a === "--days") o.days = num(argv[++i], "--days");
    else if (a === "--root") o.root = argv[++i] ?? o.root;
    else if (a === "--out") o.out = argv[++i] ?? o.out;
    else if (a === "--project") {
      const v = argv[++i];
      if (v) o.projects.push(v);
    } else if (a === "--format") {
      const v = argv[++i];
      if (v !== "md" && v !== "json") {
        console.error(`--format must be md or json, got ${String(v)}`);
        process.exit(1);
      }
      o.format = v;
    } else if (a === "--top") o.top = num(argv[++i], "--top");
    else if (a === "--no-tables") o.writeTables = false;
    else if (a === "--no-redact") o.redact = false;
    else if (a === "--pricing") {
      const v = argv[++i];
      if (!v) {
        console.error("--pricing needs a file path");
        process.exit(1);
      }
      o.pricingPath = v;
      o.pricing = loadPricing(v);
    } else if (a === "-h" || a === "--help") {
      console.error(USAGE);
      process.exit(0);
    } else {
      console.error(`Unknown argument: ${a}\n${USAGE}`);
      process.exit(1);
    }
  }
  return o;
};

const main = () => {
  const options = parseArgs(process.argv.slice(2));
  process.stderr.write(`scanning ${options.root}, last ${options.days} days\n`);

  const scan = scanTranscripts(options);
  if (scan.filesScanned === 0) {
    console.error(
      `No transcripts in the last ${options.days} days under ${options.root}. ` +
        "Check --root, widen --days, or drop --project.",
    );
  }
  if (options.writeTables) {
    const written = writeTables(scan.tables, options.out);
    process.stderr.write(`wrote ${written.length} fact tables to ${options.out}\n`);
    process.stderr.write(
      options.redact
        ? "  the tables hold real prompts and command output, redacted best-effort: review before sharing\n"
        : "  WARNING: redaction DISABLED (--no-redact): the tables hold raw prompts and secrets\n",
    );
  }

  // The aggregation reads the fact tables; the scan counters ride along on the
  // options because only the scanner can know them.
  const aggOptions: AggregateOptions = {
    ...options,
    filesScanned: scan.filesScanned,
    filesSkipped: scan.filesSkipped,
    projectCount: scan.projectCount,
    fromEpoch: scan.fromEpoch,
    toEpoch: scan.toEpoch,
  };
  const report = aggregate(scan.tables, aggOptions);
  if (options.format === "json") {
    process.stdout.write(`${JSON.stringify(report, null, 2)}\n`);
    return;
  }
  process.stdout.write(`${renderMarkdown(report, options)}\n`);
  // The work dir is only worth printing when something was written there.
  if (options.writeTables) process.stdout.write(`${options.out}\n`);
};

main();
